compiler/qsc_parse/src/prim.rs

```diff
@@ -176,6 +176,11 @@ pub(super) fn seq<T>(s: &mut Scanner, mut p: impl Parser<T>) -> Result<(Vec<T>,
     while let Some(x) = opt(s, &mut p)? {
         xs.push(x);
         if token(s, TokenKind::Comma).is_ok() {
+            while s.peek().kind == TokenKind::Comma {
```

Comment:
Oh, good call out, that's a valuable thing to consider. I think this case is a fascinating instance of the contrary: there are many positive cases and only one negative case. It turns out that the only place the comma token can legitimately appear is in a sequence parsed by `seq`, since it is literally only used for comma separated lists. So `seq` looks for e.g. `T, T, T` (with or without a final trailing comma), but handles many `T`. So if the next non-whitespace, non-comment token after a comma is a comma, the only thing that can mean is a duplicate comma in a sequence. But if we try to get the positive case, we have to ensure all the parsers of `T` that get passed into `seq` recover properly on an unexpected comma being the first token.

I think that makes this the right approach, but certainly happy to discuss and think it through more.